More unit testifications

rbtcollins opened the pull request:
See the individual commits please

rbtcollins committed:
Finish unit-testification of Windows update tests

One integration test kept as a smoke test for insurance.

rbtcollins committed:
Drop never used test

workingjubilee commented on tests/cli-self-upd.rs:
```diff
@@ -494,32 +494,25 @@ fn when_cargo_home_is_the_default_write_path_specially() {
 }
 
 #[test]
 #[cfg(unix)]
+/// This covers both windows and unix no-path-modification cases due to the code structure.
 fn install_doesnt_modify_path_if_passed_no_modify_path() {
```
This test will not compile on Windows as it is set to be `#[cfg(unix)]`, and so provides no test coverage on Windows. This is observable: It is absent from the build logs as well.

rbtcollins replied:
yes, I know. I think this is fine: we run windows and unix tests together in CI, so the test won't fail to run.

The code structure in the library means that if this test fails, we know that the platform specific do_add_to_path routines are being called inappropriately, and thats from the platform agnostic code - but because the routines that are being called are platform specific the only way we can detect this is by having some platform specific mechanism.

As this is a blackbox test, we can't drop a replacement trait in and mock it without compromising the binary that we build; a future unit test in self_update.rs might provide better coverage in a platform neutral way with a unit-test-specific trait object, but for now I think this is fine.

The trade off is that a windows developer that breaks this test will only find out in CI, which I think is ok.

workingjubilee replied:
Fair enough.

rbtcollins replied:
Oh, I should add, the equivalent platform specific test on the windows side has to modify global state - the windows registry - which is potentially disruptive, and not worth doing for an edge case test in the first place when other coverage means are available.

rbtcollins committed:
Report the toolchain desc on install

Just reporting 'stable' is less clear than we can be, and inconsistent,
since a reinstall when a default toolchain has been selected will
already report the toolchain desc (via toolchain.name()).

Also convert install_sets_up_stable into unit test.

This involves making a bunch of test support code available to unit
tests, and breaking maybe_install_rust into a calculating part that can
be tested, as well as a simple execution part which acts on the
calculated result, which we can test separately in functional tests /
tests of that underlying layer / integration tests.

this_host_triple in addition to being moved has been made somewhat more
faithful to from_host_or_build by always returning msvc for windows
hosts. Additionally, since it doesn't need to override the BUILD_TRIPLE,
that was removed from it.

kinnison commented on src/lib.rs:
```diff
@@ -66,5 +66,7 @@ pub mod fallback_settings;
 mod install;
 mod notifications;
 pub mod settings;
+// #[cfg(test)]
```
?

rbtcollins replied:
oh hmm, I think I wanted that to be cfg(test) but it can't be for the integration tests to use the test helpers, they need to be built normally and then dropped during link

kinnison reviewed (commented):
It looks consistent, modulo the -gnu failure case we need to resolve, but what's that comment all about?

kinnison reviewed (approved):
This all seems to make sense.

kinnison commented:
You are right that it took quite a kicking from some heavy boots, but I think this is good.
